nixos/modules/services/home-automation/matter-server.nix

```
@@ -0,0 +1,123 @@
+    };
+
+    port = mkOption {
+      type = types.port;
+      default = 5580;
+      defaultText = "5580";
+      description = "Port to expose the matter-server service on.";
+    };
+
+    vendorId = mkOption {
+      type = types.ints.u16;
+      default = 4939; # Home Assistant
+      defaultText = "4939 (Home Assistant)";
+      description = "CSA Vendor ID to advertise to devices on the network.";
+    };
+
+    fabricId = mkOption {
+      type = types.int;
+      default = 1;
+      defaultText = "4939 (Home Assistant)";
+      description = "CSA Vendor ID to advertise to devices on the network.";
+    };
+
+    storagePath = mkOption {
+      type = types.nonEmptyStr;
+      default = "matter-server";
+      description = mdDoc "Directory name under `/var/lib/` to store persistent data. Specifying `\"matter-server\"` indicates the directory `/var/lib/matter-server`";
+    };
+
+    logLevel = mkOption {
+      type = types.enum [ "critical" "error" "warning" "info" "debug" ];
+      default = "info";
+      description = "Verbosity of logs from the matter-server";
+    };
+
+    extraArgs = mkOption {
+      type = listOf str;
+      default = [];
+      description = "Extra arguments to pass to the matter-server executable.";
+    };
+  };
+
+  config = mkIf cfg.enable {
+    services.avahi.enable = true;
+
+    systemd.services.matter-server = {
+      after = [
+        "network-online.target"
+        "avahi-daemon.target"
+      ];
+      wantedBy = [
+        "multi-user.target"
+        "home-assistant.target"
+      ];
+      description = "Matter Server";
+      environment.HOME = "/var/lib/${cfg.storagePath}";
+      serviceConfig = {
+        ExecStart = (concatStringsSep " " [
+          "${cfg.package}/bin/matter-server"
+          "--port" (toString cfg.port)
+          "--vendorid" (toString cfg.vendorId)
+          "--fabricid" (toString cfg.fabricId)
+          "--storage-path" "/var/lib/${cfg.storagePath}"
+          "--log-level" "${cfg.logLevel}"
+          "${escapeShellArgs cfg.extraArgs}"
+        ]);
+        TemporaryFileSystem = "/data:rw /var/lib:ro";
+        BindPaths = "/var/lib/${cfg.storagePath}:/data";
+        StateDirectory = "${cfg.storagePath}";
+        ReadOnlyPaths = "/run/dbus";
+
+        # Hardening bits
+        CapabilityBoundingSet = [ "" ];
+        DevicePolicy = "closed";
+        DynamicUser = true;
+        LockPersonality = true;
+        MemoryDenyWriteExecute = true;
+        NoNewPrivileges=true;
+        PrivateDevices=true;
+        PrivateTmp = true;
+        PrivateUsers = true;
+        ProcSubset = "pid";
+        ProtectClock = true;
+        ProtectControlGroups= true;
+        ProtectHome = true;
+        ProtectHostname = true;
+        ProtectKernelLogs = true;
+        ProtectKernelModules = true;
+        ProtectKernelTunables = true;
+        ProtectProc = "invisible";
+        RestrictAddressFamilies = [ "AF_INET" "AF_INET6" "AF_UNIX" "AF_NETLINK" "AF_BLUETOOTH" ];
```

```suggestion
        RestrictAddressFamilies = [
        	"AF_INET"
        	"AF_INET6"
        	"AF_UNIX"
        	"AF_NETLINK"
        	"AF_BLUETOOTH"
        ];
```

Expanding lists makes them better readable and more robust when a merge is required.

I'm somewhat surprised by `AF_UNIX`, `AF_NETLINK` and `AF_BLUETOOTH`, since my idea of the matter-server was, that it supported either Thread or WLAN.

---

Good call-out. AF_UNIX sockets were not required, nor does it seem AF_BLUETOOTH.

Blocking AF_NETLINK sockets results in this test failure:

```
machine # [   13.109288] matter-server[951]: 2024-01-15 20:56:52 machine chip.DL[951] ERROR Failed to init netlink socket for IP addresses: 97
machine # [   13.111273] matter-server[951]: Traceback (most recent call last):
machine # [   13.112619] matter-server[951]:   File "/nix/store/h985xfmvi4dfpq51z2pa95d0k3mvvg6n-python3.11-python-matter-server-5.1.1/bin/.matter-server-wrapped", line 9, in <module>
machine # [   13.115242] matter-server[951]:     sys.exit(main())
machine # [   13.116517] matter-server[951]:              ^^^^^^
machine # [   13.117864] matter-server[951]:   File "/nix/store/h985xfmvi4dfpq51z2pa95d0k3mvvg6n-python3.11-python-matter-server-5.1.1/lib/python3.11/site-packages/matter_server/server/__main__.py", line 82, in main
machine # [   13.120922] matter-server[951]:     server = MatterServer(
machine # [   13.128743] matter-server[951]:              ^^^^^^^^^^^^^
machine # [   13.131310] matter-server[951]:   File "/nix/store/h985xfmvi4dfpq51z2pa95d0k3mvvg6n-python3.11-python-matter-server-5.1.1/lib/python3.11/site-packages/matter_server/server/server.py", line 74, in __init__
machine # [   13.138086] matter-server[951]:     self.stack = MatterStack(self)
machine # [   13.140959] matter-server[951]:                  ^^^^^^^^^^^^^^^^^
machine # [   13.144398] matter-server[951]:   File "/nix/store/h985xfmvi4dfpq51z2pa95d0k3mvvg6n-python3.11-python-matter-server-5.1.1/lib/python3.11/site-packages/matter_server/server/stack.py", line 32, in __init__
machine # [   13.152656] matter-server[951]:     self._chip_stack = ChipStack(
machine # [   13.156391] matter-server[951]:                        ^^^^^^^^^^
machine # [   13.159454] matter-server[951]:   File "/nix/store/sp558r7pil10npl538dan75gqkpw56f9-python3.11-home-assistant-chip-core-2023.12.0/lib/python3.11/site-packages/chip/ChipStack.py", line 64, in wrapper
machine # [   13.166879] matter-server[951]:     instance[0] = cls(*args, **kwargs)
machine # [   13.171054] matter-server[951]:                   ^^^^^^^^^^^^^^^^^^^^
machine # [   13.174143] matter-server[951]:   File "/nix/store/sp558r7pil10npl538dan75gqkpw56f9-python3.11-home-assistant-chip-core-2023.12.0/lib/python3.11/site-packages/chip/ChipStack.py", line 271, in __init__
machine # [   13.182048] matter-server[951]:     res.raise_on_error()
machine # [   13.184846] matter-server[951]:   File "/nix/store/sp558r7pil10npl538dan75gqkpw56f9-python3.11-home-assistant-chip-core-2023.12.0/lib/python3.11/site-packages/chip/native/__init__.py", line 67, in raise_on_error
machine # [   13.192148] matter-server[951]:     raise self.to_exception()
machine # [   13.199982] matter-server[951]: chip.exceptions.ChipStackError: src/platform/Linux/PlatformManagerImpl.cpp:165: Error 0x000000AC


```